### PR TITLE
Add TBD Impl

### PR DESCRIPTION
**Pull request by decentralgabe**
Does not yet support *-jwt variants

### CODE REVIEW

**Comment by decentralgabe** [comment]
I'm having trouble running the spruce image locally, though it passes in CI, which is why I haven't updated the report files.

**Review by OR13 — APPROVED** [review]

**Comment by OR13** [comment]
https://github.com/decentralized-identity/JWS-Test-Suite/blob/main/.github/workflows/ci.yml#L2

**Comment by decentralgabe** [comment]
All are working correctly except RSA, which I'm skeptical of. I see there is only the spruce impl passing, and it's also verifying.

**Comment by clehner** [comment]
> All are working correctly except RSA, which I'm skeptical of. I see
> there is only the spruce impl passing, and it's also verifying. 

I see that TBD is using RS256 here. Spruce is verifying only PS256 in JsonWebSignature2020. The spec says PS256. Should RS256 be allowed also?

**Comment by decentralgabe** [comment]
> All are working correctly except RSA, which I'm skeptical of. I see there is only the spruce impl passing, and it's also verifying.
> I see that TBD is using RS256 here. Spruce is verifying only PS256 in JsonWebSignature2020. The spec says PS256. Should RS256 be allowed also?

Thanks for finding this! I'll update our impl, it's a mistake.

**Comment by decentralgabe** [comment]
> > All are working correctly except RSA, which I'm skeptical of. I see there is only the spruce impl passing, and it's also verifying.
> > I see that TBD is using RS256 here. Spruce is verifying only PS256 in JsonWebSignature2020. The spec says PS256. Should RS256 be allowed also?
> 
> Thanks for finding this! I'll update our impl, it's a mistake.

@clehner appears to still be failing with PS256. I notice your implementation adds the `holder` property during signing, which is optional. I'm wondering if this is the issue, or there could be something else going on.

**Comment by decentralgabe** [comment]
looks like that fixed all but one @clehner 
https://identity.foundation/JWS-Test-Suite/implementations/tbd/presentation-0--key-4-rsa2048.vp.json

🤔

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,22 +1,28 @@
 name: CI
 on: [push, workflow_dispatch]
 jobs:
-  ci:
+  build:
```
Comment (decentralgabe) [inline .github/workflows/ci.yml:4]: @OR13 split this out into two

**Comment by clehner** [comment]
Nice!

>  I notice your implementation adds the `holder` property during signing, which is optional. I'm wondering if this is the issue, or there could be something else going on.

> https://identity.foundation/JWS-Test-Suite/implementations/tbd/presentation-0--key-4-rsa2048.vp.json

Transmute's and did-jwt-vc also add the holder property. But Transmute's verifier doesn't seem to be requiring it during verification. I thought that the holder property should be used to determine what verification methods are allowed, (analogous to the issuer property for a VC), i.e. by resolving the holder DID and seeing what the verification methods are in the resolved DID document. Therefore the holder property must be present; and since it's not in presentation-0.json, it has to be added before signing. Is this not correct? Should rather any signature/proof from the test suite's example keypair be allowed, regardless of the holder property?

**Comment by decentralgabe** [comment]
> Therefore the holder property must be present; and since it's not in presentation-0.json, it has to be added before signing. Is this not correct? Should rather any signature/proof from the test suite's example keypair be allowed, regardless of the holder property?

Your last sentence is my understanding. Since the proof block displays a verificationMethod, it is sufficient to verify the presentation. I believe this is consistent [with the specification](https://www.w3.org/TR/vc-data-model/#presentations-0) which notes "if present" for the property, which leads me to believe it's optional. If it's optional, and the proof block is present, the presentation should still be verifiable. The examples in the spec demonstrate presentations without holder properties and with proof blocks as well.

I'd be interested to hear @OR13's thoughts too.

**Comment by OR13** [comment]
VPs have only 2 required fields. `@context` and `type`.

The verification logic for VPs is by convention.

I don't think holder needs to match the verificationMethod in the proof... like how VCs require that properly.

But I could be mistaken, and I think most folks would probably assume they behave the same.

**Comment by decentralgabe** [comment]
created https://github.com/decentralized-identity/JWS-Test-Suite/issues/52

**Pull request merged**